### PR TITLE
[Impeller] create a 300 es variant of all GLES shaders to support UBO binding.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Create a GLES3 "backend" by compiling a second set of GLES shaders to 300 es. This allows the usage of UBOs and SSBOs.

### CODE REVIEW

File: shell/common/shell.cc [hunk 1]
```diff
@@ -203,8 +204,18 @@ static impeller::RuntimeStageBackend DetermineRuntimeStageBackend(
   switch (impeller_context->GetBackendType()) {
```
Comment (jonahwilliams) [inline shell/common/shell.cc:204]: This whole thing should be a method on ContentContext

File: impeller/renderer/backend/gles/device_buffer_gles.cc [hunk 2]
```diff
@@ -7,6 +7,7 @@
 #include <cstring>
 #include <memory>
 
+#include "GLES3/gl3.h"
```
Comment (chinmaygarde) [inline impeller/renderer/backend/gles/device_buffer_gles.cc:10]: Like vk.h, the convention in the GL backend is to include `//impeller/renderer/backend/gles/gles.h`.

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc [hunk 3]
```diff
@@ -101,6 +103,38 @@ bool BufferBindingsGLES::ReadUniformsBindings(const ProcTableGLES& gl,
...
+}
+
+bool BufferBindingsGLES::ReadUniformsBindingsV3(const ProcTableGLES& gl,
+                                                GLuint program) {
+  program_handle_ = program;
+  GLint uniform_blocks = 0;
+  gl.GetProgramiv(program, GL_ACTIVE_UNIFORM_BLOCKS, &uniform_blocks);
+  for (GLint i = 0; i < uniform_blocks; i++) {
+    GLint name_length = 0;
+    gl.GetActiveUniformBlockiv(program, i, GL_UNIFORM_BLOCK_NAME_LENGTH,
```
Comment (chinmaygarde) [inline impeller/renderer/backend/gles/buffer_bindings_gles.cc:120]: Paranoid note: Does the name include the NULL terminator?

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc [hunk 4]
```diff
@@ -101,6 +103,38 @@ bool BufferBindingsGLES::ReadUniformsBindings(const ProcTableGLES& gl,
...
+                               &name_length);
+
+    std::vector<GLchar> name;
+    name.resize(name_length);
+    GLint length;
```
Comment (chinmaygarde) [inline impeller/renderer/backend/gles/buffer_bindings_gles.cc:125]: Like name_length above, zero this out so an error in the next call doesn't leave garbage in the variable that you will later use to create a string.

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc [hunk 5]
```diff
@@ -101,6 +103,38 @@ bool BufferBindingsGLES::ReadUniformsBindings(const ProcTableGLES& gl,
...
+    gl.GetActiveUniformBlockName(program, i, name_length, &length, name.data());
+
+    GLuint block_index = gl.GetUniformBlockIndex(program, name.data());
+    ubo_locations_[std::string{name.data(), static_cast<size_t>(length)}] =
+        std::make_pair(block_index, i);
+  }
+  use_ubo_ = true;
+  return ReadUniformsBindingsV2(gl, program);
```
Comment (chinmaygarde) [inline impeller/renderer/backend/gles/buffer_bindings_gles.cc:133]: Nicely done.

File: impeller/renderer/backend/gles/device_buffer_gles.h [hunk 6]
```diff
@@ -34,12 +34,17 @@ class DeviceBufferGLES final
...
     kArrayBuffer,
     kElementArrayBuffer,
+    kUniformBuffer,
   };
 
   [[nodiscard]] bool BindAndUploadDataIfNecessary(BindingType type) const;
 
   void Flush(std::optional<Range> range = std::nullopt) const override;
 
+  std::optional<GLuint> GetHandle() const {
```
Comment (chinmaygarde) [inline impeller/renderer/backend/gles/device_buffer_gles.h:44]: Nit: OOL?

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc [hunk 7]
```diff
@@ -320,65 +396,44 @@ bool BufferBindingsGLES::BindUniformBuffer(const ProcTableGLES& gl,
...
+                      element_count,  // count
+                      buffer_data     // data
+        );
+        continue;
+      case sizeof(Scalar):
+        gl.Uniform1fv(location,       // location
+                      element_count,  // count
+                      buffer_data     // data
+        );
+        continue;
```
Comment (chinmaygarde) [inline impeller/renderer/backend/gles/buffer_bindings_gles.cc:436]: Perhaps add the `default:` case that logs the member type so we don't silently ignore stuff? But yeah, at this point, probably all types have shaken out.

**Review by chinmaygarde — APPROVED** [review]

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc (hunk 3 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc:120]: according to https://registry.khronos.org/OpenGL-Refpages/gl4/html/glGetActiveUniformBlock.xhtml it does

File: impeller/renderer/backend/gles/device_buffer_gles.cc (hunk 2 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/device_buffer_gles.cc]: Done

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc (hunk 4 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc]: Done

File: impeller/renderer/backend/gles/device_buffer_gles.h (hunk 6 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/device_buffer_gles.h]: Done

File: impeller/renderer/backend/gles/buffer_bindings_gles.cc (hunk 7 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/gles/buffer_bindings_gles.cc:436]: Done

**Comment by jonahwilliams** [comment]
It looks like the code size hit for this is about ~120 Kb. I don't know if this is compressed or uncompressed.

https://flutter-flutter-perf.skia.org/e/?keys=X7d49843668c5a555309193ee145055fe&selected=commit%3D43599%26name%3D%252Carch%253Darm%252Cbranch%253Dmaster%252Cconfig%253Ddefault%252Cdevice_type%253Dmokey%252Cdevice_version%253Dnone%252Chost_type%253Dmac%252Csub_result%253Dlibflutter_uncompressed_bytes%252Ctest%253Dhello_world_android__compile%252C&xbaroffset=43598

**Pull request merged**